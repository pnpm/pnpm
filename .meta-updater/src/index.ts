import fs from 'fs'
import path from 'path'
import { readWantedLockfile, Lockfile } from '@pnpm/lockfile-file'
import { ProjectManifest } from '@pnpm/types'
import { createUpdateOptions, FormatPluginFnOptions } from '@pnpm/meta-updater'
import isSubdir from 'is-subdir'
import loadJsonFile from 'load-json-file'
import normalizePath from 'normalize-path'
import exists from 'path-exists'
import writeJsonFile from 'write-json-file'

const NEXT_TAG = 'next-8'
const CLI_PKG_NAME = 'pnpm'

export default async (workspaceDir: string) => {
  const pnpmManifest = loadJsonFile.sync<any>(path.join(workspaceDir, 'pnpm/package.json'))
  const pnpmVersion = pnpmManifest!['version'] // eslint-disable-line
  const pnpmMajorKeyword = `pnpm${pnpmVersion.split('.')[0]}`
  const utilsDir = path.join(workspaceDir, '__utils__')
  const lockfile = await readWantedLockfile(workspaceDir, { ignoreIncompatible: false })
  if (lockfile == null) {
    throw new Error('no lockfile found')
  }
  return createUpdateOptions({
    'package.json': (manifest: ProjectManifest & { keywords?: string[] } | null, { dir }) => {
      if (!manifest) {
        return manifest;
      }
      if (manifest.name === 'monorepo-root') {
        manifest.scripts!['release'] = `pnpm --filter=@pnpm/exe publish --tag=${NEXT_TAG} --access=public && pnpm publish --filter=!pnpm --filter=!@pnpm/exe --access=public && pnpm publish --filter=pnpm --tag=${NEXT_TAG} --access=public`
        return manifest
      }
      if (manifest.name && manifest.name !== CLI_PKG_NAME) {
        manifest.devDependencies = {
          ...manifest.devDependencies,
          [manifest.name]: `workspace:*`,
        }
      } else if (manifest.name === CLI_PKG_NAME && manifest.devDependencies) {
        delete manifest.devDependencies[manifest.name]
      }
      if (manifest.private || isSubdir(utilsDir, dir)) return manifest
      manifest.keywords = [
        pnpmMajorKeyword,
        ...(manifest.keywords ?? []).filter((keyword) => !/^pnpm[0-9]+$/.test(keyword)),
      ]
      if (dir.includes('artifacts') || manifest.name === '@pnpm/exe') {
        manifest.version = pnpmVersion
        if (manifest.name === '@pnpm/exe') {
          for (const depName of ['@pnpm/linux-arm64', '@pnpm/linux-x64', '@pnpm/win-x64', '@pnpm/macos-x64', '@pnpm/macos-arm64']) {
            manifest.optionalDependencies![depName] = `workspace:*`
          }
        }
        return manifest
      }
      return updateManifest(workspaceDir, manifest, dir)
    },
    'tsconfig.json': updateTSConfig.bind(null, {
      lockfile,
      workspaceDir,
    }),
    'cspell.json': (cspell: any) => {
      if (cspell?.words) {
        cspell.words = cspell.words.sort()
      }
      return cspell
    },
  })
}

async function updateTSConfig (
  context: {
    lockfile: Lockfile
    workspaceDir: string
  },
  tsConfig: object | null,
  {
    dir,
    manifest,
  }: FormatPluginFnOptions
) {
  if (tsConfig == null) return tsConfig
  if (manifest.name === '@pnpm/tsconfig') return tsConfig
  const relative = normalizePath(path.relative(context.workspaceDir, dir))
  const importer = context.lockfile.importers[relative]
  if (!importer) return tsConfig
  const deps = {
    ...importer.dependencies,
    ...importer.devDependencies,
  }
  const references = [] as Array<{ path: string }>
  for (const [depName, spec] of Object.entries(deps)) {
    if (!spec.startsWith('link:') || spec.length === 5) continue
    const relativePath = spec.slice(5)
    const linkedPkgDir = path.join(dir, relativePath)
    if (!await exists(path.join(linkedPkgDir, 'tsconfig.json'))) continue
    if (!isSubdir(context.workspaceDir, linkedPkgDir)) continue
    if (
      depName === '@pnpm/package-store' && (
        manifest.name === '@pnpm/git-fetcher' ||
        manifest.name === '@pnpm/tarball-fetcher' ||
        manifest.name === '@pnpm/package-requester'
      )
    ) {
      // This is to avoid a circular graph (which TypeScript references do not support.
      continue
    }
    references.push({ path: relativePath })
  }
  await writeJsonFile(path.join(dir, 'tsconfig.lint.json'), {
    extends: './tsconfig.json',
    include: [
      'src/**/*.ts',
      'test/**/*.ts',
      '../../__typings__/**/*.d.ts',
    ],
  }, { indent: 2 })
  return {
    ...tsConfig,
    extends: '@pnpm/tsconfig',
    composite: true,
    compilerOptions: {
      ...(tsConfig as any)['compilerOptions'],
      rootDir: 'src',
    },
    references: references.sort((r1, r2) => r1.path.localeCompare(r2.path)),
  }
}

let registryMockPort = 7769

async function updateManifest (workspaceDir: string, manifest: ProjectManifest, dir: string) {
  const relative = normalizePath(path.relative(workspaceDir, dir))
  let scripts: Record<string, string>
  switch (manifest.name) {
  case '@pnpm/lockfile-types':
    scripts = { ...manifest.scripts }
    break
  case '@pnpm/headless':
  case '@pnpm/outdated':
  case '@pnpm/package-requester':
  case '@pnpm/plugin-commands-import':
  case '@pnpm/plugin-commands-installation':
  case '@pnpm/plugin-commands-listing':
  case '@pnpm/plugin-commands-outdated':
  case '@pnpm/plugin-commands-patching':
  case '@pnpm/plugin-commands-publishing':
  case '@pnpm/plugin-commands-rebuild':
  case '@pnpm/plugin-commands-script-runners':
  case '@pnpm/plugin-commands-store':
  case '@pnpm/plugin-commands-deploy':
  case CLI_PKG_NAME:
  case '@pnpm/core': {
    // @pnpm/core tests currently works only with port 4873 due to the usage of
    // the next package: pkg-with-tarball-dep-from-registry
    const port = manifest.name === '@pnpm/core' ? 4873 : ++registryMockPort
    scripts = {
      ...(manifest.scripts as Record<string, string>),
    }
    scripts.test = 'pnpm run compile && pnpm run _test'
    scripts._test = `cross-env PNPM_REGISTRY_MOCK_PORT=${port} jest`
    break
  }
  default:
    if (await exists(path.join(dir, 'test'))) {
      scripts = {
        ...(manifest.scripts as Record<string, string>),
        _test: 'jest',
        test: 'pnpm run compile && pnpm run _test',
      }
    } else {
      scripts = {
        ...(manifest.scripts as Record<string, string>),
        test: 'pnpm run compile',
      }
    }
    break
  }
  if (manifest.name === CLI_PKG_NAME) {
    manifest.publishConfig!.tag = NEXT_TAG
  }
  if (scripts._test) {
    if (scripts.pretest) {
      scripts._test = `pnpm pretest && ${scripts._test}`
    }
    if (scripts.posttest) {
      scripts._test = `${scripts._test} && pnpm posttest`
    }
    if (manifest.name === '@pnpm/server') {
      scripts._test += ' --detectOpenHandles'
    }
  }
  scripts.compile = 'tsc --build && pnpm run lint --fix'
  delete scripts.tsc
  let homepage: string
  let repository: string | { type: 'git', url: string }
  if (manifest.name === CLI_PKG_NAME) {
    homepage = 'https://pnpm.io'
    repository = {
      type: 'git',
      url: 'git+https://github.com/pnpm/pnpm.git',
    }
    scripts.compile += ' && rimraf dist bin/nodes && pnpm run bundle \
&& shx cp -r node-gyp-bin dist/node-gyp-bin \
&& shx cp -r node_modules/@pnpm/tabtab/lib/scripts dist/scripts \
&& shx cp -r node_modules/ps-list/vendor dist/vendor \
&& shx cp pnpmrc dist/pnpmrc'
  } else {
    scripts.prepublishOnly = 'pnpm run compile'
    homepage = `https://github.com/pnpm/pnpm/blob/main/${relative}#readme`
    repository = `https://github.com/pnpm/pnpm/blob/main/${relative}`
  }
  if (scripts.lint) {
    if (fs.existsSync(path.join(dir, 'test'))) {
      scripts.lint = 'eslint "src/**/*.ts" "test/**/*.ts"'
    } else {
      scripts.lint = 'eslint "src/**/*.ts"'
    }
  }
  const files: string[] = []
  if (manifest.name === CLI_PKG_NAME || manifest.name?.endsWith('/pnpm')) {
    files.push('dist')
    files.push('bin')
  } else {
    // the order is important
    files.push('lib')
    files.push('!*.map')
    if (manifest.bin) {
      files.push('bin')
    }
  }
  if (manifest.dependencies?.['@types/ramda']) {
    // We should never release @types/ramda as a prod dependency as it breaks the bit repository.
    manifest.devDependencies = {
      ...manifest.devDependencies,
      '@types/ramda': manifest.dependencies['@types/ramda'],
    }
    delete manifest.dependencies['@types/ramda']
  }
  return {
    ...manifest,
    bugs: {
      url: 'https://github.com/pnpm/pnpm/issues',
    },
    engines: {
      node: '>=16.14',
    },
    files,
    funding: 'https://opencollective.com/pnpm',
    homepage,
    license: 'MIT',
    repository,
    scripts,
    exports: {
      '.': manifest.name === 'pnpm' ? './package.json' : './lib/index.js',
    },
  }
}
