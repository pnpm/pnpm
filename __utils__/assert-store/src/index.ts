import path from 'path'
import { getFilePathInCafs } from '@pnpm/store.cafs'
import { getIntegrity, REGISTRY_MOCK_PORT } from '@pnpm/registry-mock'
import exists from 'path-exists'

export function assertStore (
  storePath: string | Promise<string>,
  encodedRegistryName?: string
) {
  // eslint-disable-next-line
  const ok = (value: any) => expect(value).toBeTruthy()
  // eslint-disable-next-line
  const notOk = (value: any) => expect(value).toBeFalsy()
  const ern = encodedRegistryName ?? `localhost+${REGISTRY_MOCK_PORT}`
  const store = {
    async getPkgIndexFilePath (pkgName: string, version?: string): Promise<string> {
      const cafsDir = path.join(await storePath, 'files')
      const integrity = version ? getIntegrity(pkgName, version) : pkgName
      return getFilePathInCafs(cafsDir, integrity, 'index')
    },
    async cafsHas (pkgName: string, version?: string): Promise<void> {
      const pathToCheck = await store.getPkgIndexFilePath(pkgName, version)
      ok(await exists(pathToCheck))
    },
    async cafsHasNot (pkgName: string, version?: string): Promise<void> {
      const pathToCheck = await store.getPkgIndexFilePath(pkgName, version)
      notOk(await exists(pathToCheck))
    },
    async storeHas (pkgName: string, version?: string): Promise<void> {
      const pathToCheck = await store.resolve(pkgName, version)
      ok(await exists(pathToCheck))
    },
    async storeHasNot (pkgName: string, version?: string): Promise<void> {
      const pathToCheck = await store.resolve(pkgName, version)
      notOk(await exists(pathToCheck))
    },
    async resolve (pkgName: string, version?: string, relativePath?: string): Promise<string> {
      const pkgFolder = version ? path.join(ern, pkgName, version) : pkgName
      if (relativePath) {
        return path.join(await storePath, pkgFolder, relativePath)
      }
      return path.join(await storePath, pkgFolder)
    },
  }
  return store
}
