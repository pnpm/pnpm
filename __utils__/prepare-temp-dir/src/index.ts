import fs from 'node:fs'
import path from 'node:path'

// The testing folder should be outside of the project to avoid lookup in the project's node_modules
// Not using the OS temp directory due to issues on Windows CI.
const tmpBaseDir = path.join(__dirname, '../../../../pnpm_tmp')

const tmpPath = path.join(tmpBaseDir, `${getFilesCountInDir(tmpBaseDir).toString()}_${process.pid.toString()}`)

let dirNumber = 0

export function tempDir (chdir: boolean = true): string {
  dirNumber++
  const dirname = dirNumber.toString()
  const tmpDir = path.join(tmpPath, dirname)
  fs.mkdirSync(tmpDir, { recursive: true })

  if (chdir) process.chdir(tmpDir)

  return tmpDir
}

function getFilesCountInDir (dir: string): number {
  try {
    return fs.readdirSync(dir).length
  } catch {
    return 0
  }
}
