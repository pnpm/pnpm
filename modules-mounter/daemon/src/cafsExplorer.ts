import type { PackageFilesIndex } from '@pnpm/types'

export function readdir(index: PackageFilesIndex, dir: string): string[] {
  const dirs = new Set<string>()

  const prefix = dir ? `${dir}/` : ''

  for (const filePath of Object.keys(index.files)) {
    if (filePath.startsWith(prefix)) {
      const parts = filePath.substring(dir.length).split('/')

      const part = parts[0] ?? parts[1]

      if (part) {
        dirs.add(part)
      }
    }
  }

  return Array.from(dirs)
}

export function dirEntityType(index: PackageFilesIndex, p: string): 'file' | 'directory' | undefined {
  if (index.files[p]) {
    return 'file'
  }

  const prefix = `${p}/`

  return Object.keys(index.files).some((k) => k.startsWith(prefix))
    ? 'directory'
    : undefined
}
