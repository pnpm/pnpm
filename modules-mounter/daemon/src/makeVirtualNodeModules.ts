import path from 'path'
import { type Lockfile } from '@pnpm/lockfile-file'
import { DEPENDENCIES_FIELDS } from '@pnpm/types'
import { nameVerFromPkgSnapshot } from '@pnpm/lockfile-utils'
import * as dp from '@pnpm/dependency-path'
import normalize from 'normalize-path'

interface DirDirEntry {
  entryType: 'directory'
  entries: Record<string, DirEntry>
}

type DirEntry = {
  entryType: 'index'
  depPath: string
} | {
  entryType: 'symlink'
  target: string
} | DirDirEntry

export function makeVirtualNodeModules (lockfile: Lockfile): DirEntry {
  const entries: Record<string, DirEntry> = {
    '.pnpm': {
      entryType: 'directory',
      entries: createVirtualStoreDir(lockfile),
    },
  }
  for (const depType of DEPENDENCIES_FIELDS) {
    for (const [depName, ref] of Object.entries(lockfile.importers['.'][depType] ?? {})) {
      const symlink: DirEntry = {
        entryType: 'symlink',
        target: `./.pnpm/${dp.depPathToFilename(dp.refToRelative(ref, depName)!, 120)}/node_modules/${depName}`,
      }
      addDirEntry(entries, depName, symlink)
    }
  }
  return {
    entryType: 'directory',
    entries,
  }
}

function createVirtualStoreDir (lockfile: Lockfile): Record<string, DirEntry> {
  const rootDir = {} as Record<string, DirEntry>
  for (const [depPath, pkgSnapshot] of Object.entries(lockfile.packages ?? {})) {
    const { name } = nameVerFromPkgSnapshot(depPath, pkgSnapshot)
    const pkgNodeModules = {} as Record<string, DirEntry>
    const currentPath = dp.depPathToFilename(depPath, 120)
    const pkgDir: DirEntry = {
      entryType: 'index',
      depPath,
    }
    addDirEntry(pkgNodeModules, name, pkgDir)
    for (const [depName, ref] of Object.entries({ ...pkgSnapshot.dependencies, ...pkgSnapshot.optionalDependencies })) {
      const symlink: DirEntry = {
        entryType: 'symlink',
        target: normalize(path.relative(`${currentPath}/node_modules/`, `${dp.depPathToFilename(dp.refToRelative(ref, depName)!, 120)}/node_modules/${depName}`)),
      }
      addDirEntry(pkgNodeModules, depName, symlink)
    }
    addDirEntry(rootDir, currentPath, {
      entryType: 'directory',
      entries: {
        node_modules: {
          entryType: 'directory',
          entries: pkgNodeModules,
        },
      },
    })
  }
  return rootDir
}

function addDirEntry (target: Record<string, DirEntry>, subPath: string[] | string, newEntry: DirEntry): void {
  const subPathArray = typeof subPath === 'string' ? subPath.split('/') : subPath
  const p = subPathArray.shift()!
  if (subPathArray.length > 0) {
    if (!target[p]) {
      target[p] = {
        entryType: 'directory',
        entries: {},
      } as DirEntry
    } else if (target[p].entryType !== 'directory') {
      throw new Error()
    }
    addDirEntry((target[p] as DirDirEntry).entries, subPathArray, newEntry)
  } else {
    target[p] = newEntry
  }
}
