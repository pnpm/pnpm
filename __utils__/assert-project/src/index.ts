import path from 'path'
import { assertStore } from '@pnpm/assert-store'
import { WANTED_LOCKFILE } from '@pnpm/constants'
import { type LockfileV6 as Lockfile, type ProjectSnapshotV6 as ProjectSnapshot } from '@pnpm/lockfile-types'
import { type Modules, readModulesManifest } from '@pnpm/modules-yaml'
import { REGISTRY_MOCK_PORT } from '@pnpm/registry-mock'
import readYamlFile from 'read-yaml-file'
import exists from 'path-exists'
import writePkg from 'write-pkg'
import isExecutable from './isExecutable'

export { isExecutable, type Modules }

export type RawLockfile = Lockfile & Partial<ProjectSnapshot>

export interface Project {
  // eslint-disable-next-line
  requireModule: (moduleName: string) => any
  dir: () => string
  has: (pkgName: string, modulesDir?: string) => Promise<void>
  hasNot: (pkgName: string, modulesDir?: string) => Promise<void>
  getStorePath: () => Promise<string>
  resolve: (pkgName: string, version?: string, relativePath?: string) => Promise<string>
  getPkgIndexFilePath: (pkgName: string, version?: string) => Promise<string>
  cafsHas: (pkgName: string, version?: string) => Promise<void>
  cafsHasNot: (pkgName: string, version?: string) => Promise<void>
  storeHas: (pkgName: string, version?: string) => Promise<string>
  storeHasNot: (pkgName: string, version?: string) => Promise<void>
  isExecutable: (pathToExe: string) => Promise<void>
  /**
   * TODO: Remove the `Required<T>` cast.
   *
   * https://github.com/microsoft/TypeScript/pull/32695 might help with this.
   */
  readCurrentLockfile: () => Promise<Required<RawLockfile>>
  readModulesManifest: () => Promise<Modules | null>
  /**
   * TODO: Remove the `Required<T>` cast.
   *
   * https://github.com/microsoft/TypeScript/pull/32695 might help with this.
   */
  readLockfile: (lockfileName?: string) => Promise<Required<RawLockfile>>
  writePackageJson: (pkgJson: object) => Promise<void>
}

export function assertProject (projectPath: string, encodedRegistryName?: string): Project {
  const ern = encodedRegistryName ?? `localhost+${REGISTRY_MOCK_PORT}`
  const modules = path.join(projectPath, 'node_modules')

  let cachedStore: {
    storePath: string
    getPkgIndexFilePath: (pkgName: string, version?: string) => Promise<string>
    cafsHas: (pkgName: string, version?: string | undefined) => Promise<void>
    cafsHasNot: (pkgName: string, version?: string | undefined) => Promise<void>
    storeHas: (pkgName: string, version?: string | undefined) => Promise<void>
    storeHasNot: (pkgName: string, version?: string | undefined) => Promise<void>
    resolve: (pkgName: string, version?: string | undefined, relativePath?: string | undefined) => Promise<string>
  }
  async function getStoreInstance () {
    if (!cachedStore) {
      const modulesYaml = await readModulesManifest(modules)
      if (modulesYaml == null) {
        throw new Error(`Cannot find module store. No .modules.yaml found at "${modules}"`)
      }
      const storePath = modulesYaml.storeDir
      cachedStore = {
        storePath,
        ...assertStore(storePath, ern),
      }
    }
    return cachedStore
  }
  async function getVirtualStoreDir () {
    const modulesYaml = await readModulesManifest(modules)
    if (modulesYaml == null) {
      return path.join(modules, '.pnpm')
    }
    return modulesYaml.virtualStoreDir
  }

  // eslint-disable-next-line
  const ok = (value: any) => expect(value).toBeTruthy()
  // eslint-disable-next-line
  const notOk = (value: any) => expect(value).toBeFalsy()
  return {
    dir: () => projectPath,
    requireModule (pkgName: string) {
      // eslint-disable-next-line
      return require(path.join(modules, pkgName))
    },
    async has (pkgName: string, _modulesDir?: string) {
      const md = _modulesDir ? path.join(projectPath, _modulesDir) : modules
      ok(await exists(path.join(md, pkgName)))
    },
    async hasNot (pkgName: string, _modulesDir?: string) {
      const md = _modulesDir ? path.join(projectPath, _modulesDir) : modules
      notOk(await exists(path.join(md, pkgName)))
    },
    async getStorePath () {
      const store = await getStoreInstance()
      return store.storePath
    },
    async resolve (pkgName: string, version?: string, relativePath?: string) {
      const store = await getStoreInstance()
      return store.resolve(pkgName, version, relativePath)
    },
    async getPkgIndexFilePath (pkgName: string, version?: string): Promise<string> {
      const store = await getStoreInstance()
      return store.getPkgIndexFilePath(pkgName, version)
    },
    async cafsHas (pkgName: string, version?: string) {
      const store = await getStoreInstance()
      return store.cafsHas(pkgName, version)
    },
    async cafsHasNot (pkgName: string, version?: string) {
      const store = await getStoreInstance()
      return store.cafsHasNot(pkgName, version)
    },
    async storeHas (pkgName: string, version?: string) {
      const store = await getStoreInstance()
      return store.resolve(pkgName, version)
    },
    async storeHasNot (pkgName: string, version?: string) {
      try {
        const store = await getStoreInstance()
        return store.storeHasNot(pkgName, version)
      } catch (err: any) { // eslint-disable-line
        if (err.message.startsWith('Cannot find module store')) {
          return
        }
        throw err
      }
    },
    async isExecutable (pathToExe: string) {
      return isExecutable(ok, path.join(modules, pathToExe))
    },
    async readCurrentLockfile () {
      try {
        return await readYamlFile(path.join(await getVirtualStoreDir(), 'lock.yaml'))
      } catch (err: any) { // eslint-disable-line
        if (err.code === 'ENOENT') return null!
        throw err
      }
    },
    readModulesManifest: async () => readModulesManifest(modules),
    async readLockfile (lockfileName: string = WANTED_LOCKFILE) {
      try {
        return await readYamlFile(path.join(projectPath, lockfileName))
      } catch (err: any) { // eslint-disable-line
        if (err.code === 'ENOENT') return null!
        throw err
      }
    },
    async writePackageJson (pkgJson: object) {
      return writePkg(projectPath, pkgJson as any) // eslint-disable-line
    },
  }
}
