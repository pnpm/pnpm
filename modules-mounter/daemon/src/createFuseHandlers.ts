// cspell:ignore ents
import fs from 'fs'
import { getFilePathInCafs, getFilePathByModeInCafs, type PackageFilesIndex } from '@pnpm/store.cafs'
import { type Lockfile, readWantedLockfile, type PackageSnapshot, type TarballResolution } from '@pnpm/lockfile-file'
import {
  nameVerFromPkgSnapshot,
} from '@pnpm/lockfile-utils'
import * as schemas from 'hyperdrive-schemas'
import loadJsonFile from 'load-json-file'
import Fuse from 'fuse-native'
import * as cafsExplorer from './cafsExplorer'
import { makeVirtualNodeModules } from './makeVirtualNodeModules'

const TIME = new Date()
const STAT_DEFAULT = {
  mtime: TIME,
  atime: TIME,
  ctime: TIME,
  nlink: 1,
  uid: process.getuid ? process.getuid() : 0,
  gid: process.getgid ? process.getgid() : 0,
}

export interface FuseHandlers {
  open: (p: string, flags: string | number, cb: (exitCode: number, fd?: number) => void) => void
  release: (p: string, fd: number, cb: (exitCode: number) => void) => void
  read: (p: string, fd: number, buffer: Buffer, length: number, position: number, cb: (readBytes: number) => void) => void
  readlink: (p: string, cb: (returnCode: number, target?: string) => void) => void
  /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
  getattr: (p: string, cb: (returnCode: number, files?: any) => void) => void
  readdir: (p: string, cb: (returnCode: number, files?: string[]) => void) => void
}

export async function createFuseHandlers (lockfileDir: string, cafsDir: string): Promise<FuseHandlers> {
  const lockfile = await readWantedLockfile(lockfileDir, { ignoreIncompatible: true })
  if (lockfile == null) throw new Error('Cannot generate a .pnp.cjs without a lockfile')
  return createFuseHandlersFromLockfile(lockfile, cafsDir)
}

export function createFuseHandlersFromLockfile (lockfile: Lockfile, cafsDir: string) {
  const pkgSnapshotCache = new Map<string, { name: string, version: string, pkgSnapshot: PackageSnapshot, index: PackageFilesIndex }>()
  const virtualNodeModules = makeVirtualNodeModules(lockfile)
  return {
    open (p: string, flags: string | number, cb: (exitCode: number, fd?: number) => void) {
      const dirEnt = getDirEnt(p)
      if (dirEnt?.entryType !== 'index') {
        cb(-1)
        return
      }
      const fileInfo = dirEnt.index.files[dirEnt.subPath]
      if (!fileInfo) {
        cb(-1)
        return
      }
      const filePathInStore = getFilePathByModeInCafs(cafsDir, fileInfo.integrity, fileInfo.mode)
      fs.open(filePathInStore, flags, (err, fd) => {
        if (err != null) {
          cb(-1)
          return
        }
        // eslint-disable-next-line n/no-callback-literal
        cb(0, fd)
      })
    },
    release (p: string, fd: number, cb: (exitCode: number) => void) {
      fs.close(fd, (err) => {
        cb((err != null) ? -1 : 0)
      })
    },
    read (p: string, fd: number, buffer: Buffer, length: number, position: number, cb: (readBytes: number) => void) {
      fs.read(fd, buffer, 0, length, position, (err, bytesRead) => {
        if (err != null) {
          cb(-1)
          return
        }
        cb(bytesRead)
      })
    },
    readlink (p: string, cb: (returnCode: number, target?: string) => void) {
      const dirEnt = getDirEnt(p)
      if (dirEnt?.entryType !== 'symlink') {
        cb(Fuse.ENOENT)
        return
      }
      // eslint-disable-next-line n/no-callback-literal
      cb(0, dirEnt.target)
    },
    /* eslint-disable-next-line @typescript-eslint/no-explicit-any */
    getattr (p: string, cb: (returnCode: number, files?: any) => void) {
      const dirEnt = getDirEnt(p)
      if (dirEnt == null) {
        cb(Fuse.ENOENT)
        return
      }
      if (dirEnt.entryType === 'directory' || dirEnt.entryType === 'index' && !dirEnt.subPath) {
        // eslint-disable-next-line n/no-callback-literal
        cb(0, schemas.Stat.directory({
          ...STAT_DEFAULT,
          size: 1,
        }))
        return
      }
      if (dirEnt.entryType === 'symlink') {
        // eslint-disable-next-line n/no-callback-literal
        cb(0, schemas.Stat.symlink({
          ...STAT_DEFAULT,
          size: 1,
        }))
        return
      }
      if (dirEnt.entryType === 'index') {
        switch (cafsExplorer.dirEntityType(dirEnt.index, dirEnt.subPath)) {
        case 'file': {
          const { size, mode } = dirEnt.index.files[dirEnt.subPath]
          // eslint-disable-next-line n/no-callback-literal
          cb(0, schemas.Stat.file({
            ...STAT_DEFAULT,
            mode,
            size,
          }))
          return
        }
        case 'directory':
        // eslint-disable-next-line n/no-callback-literal
          cb(0, schemas.Stat.directory({
            ...STAT_DEFAULT,
            size: 1,
          }))
          return
        default:
          cb(Fuse.ENOENT)
          return
        }
      }
      cb(Fuse.ENOENT)
    },
    readdir,
  }
  function readdir (p: string, cb: (returnCode: number, files?: string[]) => void) {
    const dirEnt = getDirEnt(p)
    if (dirEnt?.entryType === 'index') {
      const dirEnts = cafsExplorer.readdir(dirEnt.index, dirEnt.subPath)
      if (dirEnts.length === 0) {
        cb(Fuse.ENOENT)
        return
      }
      // eslint-disable-next-line n/no-callback-literal
      cb(0, dirEnts)
      return
    }
    if ((dirEnt == null) || dirEnt.entryType !== 'directory') {
      cb(Fuse.ENOENT)
      return
    }
    // eslint-disable-next-line n/no-callback-literal
    cb(0, Object.keys(dirEnt.entries))
  }
  function getDirEnt (p: string) {
    let currentDirEntry = virtualNodeModules
    const parts = p === '/' ? [] : p.split('/')
    parts.shift()
    while ((parts.length > 0) && currentDirEntry && currentDirEntry.entryType === 'directory') {
      currentDirEntry = currentDirEntry.entries[parts.shift()!]
    }
    if (currentDirEntry?.entryType === 'index') {
      const pkg = getPkgInfo(currentDirEntry.depPath, cafsDir)
      if (pkg == null) {
        return null
      }
      return {
        ...currentDirEntry,
        index: pkg.index,
        subPath: parts.join('/'),
      }
    }
    return currentDirEntry
  }
  function getPkgInfo (depPath: string, cafsDir: string) {
    if (!pkgSnapshotCache.has(depPath)) {
      const pkgSnapshot = lockfile.packages?.[depPath]
      if (pkgSnapshot == null) return undefined
      const indexPath = getFilePathInCafs(cafsDir, (pkgSnapshot.resolution as TarballResolution).integrity!, 'index')
      pkgSnapshotCache.set(depPath, {
        ...nameVerFromPkgSnapshot(depPath, pkgSnapshot),
        pkgSnapshot,
        index: loadJsonFile.sync<PackageFilesIndex>(indexPath), // TODO: maybe make it async?
      })
    }
    return pkgSnapshotCache.get(depPath)
  }
}
