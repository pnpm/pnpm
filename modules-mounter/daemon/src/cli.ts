import '@total-typescript/ts-reset'

import path from 'node:path'
import { promises as fs } from 'node:fs'

import { getConfig } from '@pnpm/config'
import { getStorePath } from '@pnpm/store-path'

import Fuse from 'fuse-native'

import { createFuseHandlers } from './createFuseHandlers'

;(async (): Promise<void> => {
  const mnt = path.join(process.cwd(), 'node_modules')

  await fs.mkdir(mnt, { recursive: true })

  const { config } = await getConfig({
    cliOptions: {},
    packageManager: { name: '', version: '' },
  })

  const cafsDir = path.join(
    await getStorePath({
      pkgRoot: process.cwd(),
      storePath: config.storeDir,
      pnpmHomeDir: config.pnpmHomeDir,
    }),
    'files'
  )

  const fuse = new Fuse(mnt, await createFuseHandlers(process.cwd(), cafsDir), {
    debug: true,
  })

  fuse.mount((err?: Error | undefined): void => {
    if (err != null) console.error(err)
  })

  process.once('SIGINT', (): void => {
    fuse.unmount((err?: Error | undefined) => {
      if (typeof err === 'undefined') {
        console.log(`filesystem at ${fuse.mnt as string} unmounted`)
      } else {
        console.log(`filesystem at ${fuse.mnt as string} not unmounted`, err)
      }
    })
  })
})()
