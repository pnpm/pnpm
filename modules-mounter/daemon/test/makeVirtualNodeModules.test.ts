import path from 'node:path'
import { readWantedLockfile } from '@pnpm/lockfile-file'
import { makeVirtualNodeModules } from '../src/makeVirtualNodeModules'

test('makeVirtualNodeModules', async () => {
  const lockfile = await readWantedLockfile(
    path.join(__dirname, '__fixtures__/simple'),
    { ignoreIncompatible: true }
  )
  expect(lockfile).not.toBeNull()
  // @ts-ignore
  expect(makeVirtualNodeModules(lockfile)).toMatchSnapshot()
})
