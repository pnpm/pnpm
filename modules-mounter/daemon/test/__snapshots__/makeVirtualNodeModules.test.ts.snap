// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`makeVirtualNodeModules 1`] = `
{
  "entries": {
    ".pnpm": {
      "entries": {
        "@zkochan+git-config@0.1.0": {
          "entries": {
            "node_modules": {
              "entries": {
                "@zkochan": {
                  "entries": {
                    "git-config": {
                      "depPath": "/@zkochan/git-config@0.1.0",
                      "entryType": "index",
                    },
                  },
                  "entryType": "directory",
                },
                "ini": {
                  "entryType": "symlink",
                  "target": "../../ini@1.3.4/node_modules/ini",
                },
              },
              "entryType": "directory",
            },
          },
          "entryType": "directory",
        },
        "ini@1.3.4": {
          "entries": {
            "node_modules": {
              "entries": {
                "ini": {
                  "depPath": "/ini@1.3.4",
                  "entryType": "index",
                },
              },
              "entryType": "directory",
            },
          },
          "entryType": "directory",
        },
        "is-positive@1.0.0": {
          "entries": {
            "node_modules": {
              "entries": {
                "is-positive": {
                  "depPath": "/is-positive@1.0.0",
                  "entryType": "index",
                },
              },
              "entryType": "directory",
            },
          },
          "entryType": "directory",
        },
      },
      "entryType": "directory",
    },
    "@zkochan": {
      "entries": {
        "git-config": {
          "entryType": "symlink",
          "target": "./.pnpm/@zkochan+git-config@0.1.0/node_modules/@zkochan/git-config",
        },
      },
      "entryType": "directory",
    },
    "is-positive": {
      "entryType": "symlink",
      "target": "./.pnpm/is-positive@1.0.0/node_modules/is-positive",
    },
  },
  "entryType": "directory",
}
`;
